sklearn/neighbors/nca.py

```diff
@@ -0,0 +1,545 @@
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The training samples.
+
+        y : array-like, shape (n_samples,)
+            The corresponding training labels.
+
+        Returns
+        -------
+        self : object
+            returns a trained NeighborhoodComponentAnalysis model.
+        """
+
+        # Verify inputs X and y and NCA parameters, and transform a copy if
+        # needed
+        X_valid, y_valid, init = self._validate_params(X, y)
+
+        # Initialize the random generator
+        self.random_state_ = check_random_state(self.random_state)
+
+        # Measure the total training time
+        t_train = time.time()
+
+        # Compute arrays that stay fixed during optimization:
+        # mask for fast lookup of same-class samples
+        masks = _make_masks(y_valid)
+        # pairwise differences
+        diffs = X_valid[:, np.newaxis] - X_valid[np.newaxis]
+
+        # Initialize the transformation
+        transformation = self._initialize(X_valid, init)
+
+        # Create a dictionary of parameters to be passed to the optimizer
+        disp = self.verbose - 2 if self.verbose > 1 else -1
+        optimizer_params = {'method': 'L-BFGS-B',
+                            'fun': self._loss_grad_lbfgs,
+                            'args': (X_valid, y_valid, diffs, masks),
+                            'jac': True,
+                            'x0': transformation,
+                            'tol': self.tol,
+                            'options': dict(maxiter=self.max_iter, disp=disp),
+                            'callback': self._callback
+                            }
+
+        # Call the optimizer
+        self.n_iter_ = 0
+        opt_result = minimize(**optimizer_params)
+
+        # Reshape the solution found by the optimizer
+        self.transformation_ = opt_result.x.reshape(-1, X_valid.shape[1])
+
+        # Stop timer
+        t_train = time.time() - t_train
+        if self.verbose:
+            print('[{}] Training took {:8.2f}s.'.format(
+                self.__class__.__name__, t_train))
+
+        # Optionally store information returned by the optimizer
+        if self.store_opt_result:
+            self.opt_result_ = opt_result
+
+        return self
+
+    def transform(self, X):
+        """Applies the learned transformation to the given data.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Data samples.
+
+        Returns
+        -------
+        X_embedded: array, shape (n_samples, n_features_out)
+            The data samples transformed.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        check_is_fitted(self, ['transformation_'])
+        X = check_array(X)
+
+        return np.dot(X, self.transformation_.T)
+
+    def _validate_params(self, X, y):
```

Comment:
some questions on parameter checks:
- are we robust to the case where there is only one class? In this case there is nothing to do (the objective function is 0) so we could rule out this case when checking parameters
- are we robust to singleton classes?

Reply:
I added a test for only one class and singleton classes, and they both pass. For one class I checked that if init='identity', the transformed X will be the same as X, and for singleton classes I just tested that "fit" does not throw any error (took the test from lmnn, except that lmnn should throw an error in the "one non-singleton" case)

Reply:
you could also test that when all classes are singletons, that the transformed X is also the same as X

Reply:
You are right. I added the test, and it passed.